pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -159,13 +169,67 @@ func (s *Scheduler) add(pod *v1.Pod) error {
 	// Create new node
 	var errs error
 	for _, nodeTemplate := range s.nodeTemplates {
-		node := NewNode(nodeTemplate, s.topology, s.daemonOverhead[nodeTemplate], s.instanceTypes)
+		instanceTypes := s.instanceTypes
+		// if limits have been applied to the provisioner, ensure we filter instance types to avoid violating those limits
+		if remaining, ok := s.remainingResources[nodeTemplate.ProvisionerName]; ok {
+			instanceTypes = filterByRemainingResources(s.instanceTypes, remaining)
+			if len(instanceTypes) == 0 {
+				errs = multierr.Append(errs, fmt.Errorf("all available instance types exceed provisioner limits"))
+				continue
+			}
+		}
+
+		node := NewNode(nodeTemplate, s.topology, s.daemonOverhead[nodeTemplate], instanceTypes)
 		err := node.Add(pod)
 		if err == nil {
 			s.nodes = append(s.nodes, node)
+			// we will launch this node and need to track its maximum possible resource usage against our remaining resources
+			s.remainingResources[nodeTemplate.ProvisionerName] = subtractMax(s.remainingResources[nodeTemplate.ProvisionerName], node.InstanceTypeOptions)
 			return nil
 		}
 		errs = multierr.Append(errs, err)
 	}
 	return errs
 }
+
+// subtractMax returns the remaining resources after subtracting the max resource quantity per instance type. To avoid
```

Comment:
This could cause us to require multiple scheduling rounds as you approach the limits configured on a provisioner. We still launch the max node sizes possible, but to avoid overshooting we have to assume worst case.